linkerd/app/core/src/errors.rs

```diff
@@ -217,31 +219,60 @@ impl<RspB: Default + hyper::body::HttpBody> respond::Respond<http::Response<RspB
                     return Ok(rsp);
                 }
 
-                let status = http_status(&*error);
-                debug!(%status, version = ?self.version, "Handling error with HTTP response");
-                Ok(http::Response::builder()
+                let mut rsp = http::Response::builder()
                     .version(self.version)
-                    .status(status)
                     .header(http::header::CONTENT_LENGTH, "0")
                     .body(ResponseBody::default())
-                    .expect("error response must be valid"))
+                    .expect("error response must be valid");
+                let status = set_http_status(&*error, rsp.headers_mut());
+                *rsp.status_mut() = status;
```

Comment:
hmm, it's a little weird/inconsistent to me that `set_http_status` _mutates_ the headers, but _returns_ a status which we then set as the response's status. what if, instead of taking a `&mut HeaderMap`, `set_http_status` took an `&mut http::response::Builder` instead, and set the headers _and_ the status code on the response in one place, before we completed building the response?

Reply:
Yep that makes sense. `set_http_status` now makes the `http:response::Builder` and returns it after setting the header and status that correspond to the error the proxy encountered.